src/ndb.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+#include "ndb.h"
+
+#include "ndb_p1_section.h"
+#include "ndb_p3_section.h"
+
+ndb_file_version_t ndb_file_version = 0x00000001;
+
+#define NDB_MAX_SECTIONS_NUM 2
+static ndb_section_if_t *ndb_sections[NDB_MAX_SECTIONS_NUM];
+
+
+enum ndb_op_code ndb_init()
+{
+    enum ndb_op_code op_code;
+    op_code = ndb_core_init();
+    if(NDB_ERR_NONE != op_code) {
+        return op_code;
+    }
+
+    int i;
+    for(i = 0; i < NDB_MAX_SECTIONS_NUM; i++)
+    {
+        ndb_sections[i] = NULL;
+    }
+
+    /*
+     * Platform specific code
+     */
+    if(ndb_register_section(&p1_section_if) != NDB_ERR_NONE) {
+        /*TODO: report failure */
+    }
+    if(ndb_register_section(&p3_section_if) != NDB_ERR_NONE) {
+        /*TODO: report failure */
+    }
+
+    return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_register_section(ndb_section_if_t* section_if)
+{
+    int i;
+    for(i = 0; i < NDB_MAX_SECTIONS_NUM; i++)
+    {
+        if(NULL == ndb_sections[i]) {
+            if(NDB_ERR_NONE == section_if->init()) {
+                ndb_sections[i] = section_if;
+                return NDB_ERR_NONE;
+            }
+            return NDB_ERR_SECTION_INIT_FAILURE;
+        }
+    }
+    return NDB_ERR_NO_SECTION_SPACE;
+}
+
+enum ndb_op_code ndb_process_req(ndb_request_info_t *r, size_t f_offset)
```

Comment:
I think the core module needs to be aware of (or at least be able to find) the registered data elements for purposed of writing to NVM, but I don't really see a reason why the read/write functions need to pass from core -> section-specific -> core. Why not just go straight from section-specific -> core?